PRIOR-ARTS.md

```diff
@@ -0,0 +1,547 @@
+- Ruby [Context](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb), based on Ruby's [Fiber](#ruby).
+- Rust [Context](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry/src/context.rs), does not support tokio yet.
+- Swift:
+  - [ActivityContextManager](https://github.com/open-telemetry/opentelemetry-swift/blob/main/Sources/OpenTelemetryApi/Context/ActivityContextManager.swift).
+  - [TaskLocalContextManager](https://github.com/open-telemetry/opentelemetry-swift/blob/main/Sources/OpenTelemetryApi/Context/TaskLocalContextManager.swift).
+
+## JavaScript prior arts
+
+### Node.js AsyncLocalStorage
+
+Node.js provides a stable API [`AsyncLocalStorage`][] that supports implicit context propagation
+across `await` and runtime APIs.
+
+```typescript
+class AsyncLocalStorage<ValueType> {
+  static bind<T extends Function>(fn: T): T;
+  static snapshot(): () => void;
+
+  constructor();
+
+  getStore(): ValueType;
+
+  run<T extends Function, ReturnType = GetReturnType<T>>(store: ValueType, callback: T, ...args: never[]): ReturnType;
+
+  /** @experimental */
+  enterWith(store: ValueType);
+}
+```
+
+The `AsyncContext.Variable` is significantly inspired by `AsyncLocalStorage`. However,
+`AsyncContext.Variable` only provides an essential subset of `AsyncLocalStorage`,
+with a follow-up extension for set semantic with scope enforcement
+like `using _ = asyncVar.withValue(val)`, as described in
+[mutation-scope.md](./MUTATION-SCOPE.md#the-set-semantic-with-scope-enforcement).
+
+Additionally, as `AsyncContext.Variable` is built in the language, it also
+support language constructs like (async) generators.
+
+### zones.js
+
+[`zone.js`][] provides a `Zone` object, which has the following API:
+
+```typescript
+class Zone {
+  constructor({ name, parent });
+
+  name;
+  get parent();
+
+  fork({ name });
+  run(callback);
+  wrap(callback);
+
+  static get current();
+}
+```
+
+The concept of the _current zone_, reified as `Zone.current`, is crucial. Both
+`run` and `wrap` are designed to manage running the current zone:
+
+- `z.run(callback)` will set the current zone to `z` for the duration of
+  `callback`, resetting it to its previous value afterward. This is how you
+  "enter" a zone.
+- `z.wrap(callback)` produces a new function that essentially performs
+  `z.run(callback)` (passing along arguments and this, of course).
+
+The _current zone_ is the async context that propagates with all our operations.
+In our above example, sites `(1)` through `(6)` would all have the same value of
+`Zone.current`. If a developer had done something like:
+
+```typescript
+const loadZone = Zone.current.fork({ name: "loading zone" });
+window.onload = loadZone.wrap(e => { ... });
+```
+
+then at all those sites, `Zone.current` would be equal to `loadZone`.
+
```

Comment:
Zone.js also allows doing a lot of stuff not in AsyncContext, such as allowing overriding the callback to run code before and after it. We should mention that that was not in scope for this proposal.